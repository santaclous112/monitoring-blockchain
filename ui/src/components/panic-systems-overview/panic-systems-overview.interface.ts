import {Metrics} from "../../interfaces/metrics";

export interface PanicSystemsOverviewInterface {

    /**
     * The name of the selected base chain.
     */
    baseChainName: string,

    /**
     * Array of system metrics for the given base chain name available from the
     * API.
     */
    systemMetrics: Metrics[],

    /**
     * The {@link window.setInterval} ID, necessary to clear the 'time
     * interval' once the component is 'destroyed' (detached from the DOM).
     */
    _updater: number,

    /**
     * How frequent the data is fetched (in milliseconds).
     */
    _updateFrequency: number,
}
