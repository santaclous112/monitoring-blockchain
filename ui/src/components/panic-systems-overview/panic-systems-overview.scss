.panic-systems-overview {
  &__subtitle {
    margin-block-end: 0.5em;
    color: var(--ion-color-primary);
  }

  &__info-message {
    padding-left: 5px;
    padding-top: 10px;
  }
}

panic-systems-overview {
  svc-surface > div > svc-label {
    font-weight: bold;
    font-size: 25px;
    margin: 30px 0 30px;
    color: var(--ion-color-primary);
  }

  svc-data-table tr {
    td.info {
      color: #4CBB17;
    }

    td.warning {
      color: #FFA500;
    }

    td.critical {
      color: #FF0000;
    }

    td.not-available {
      color: #808080;
    }
  }
}
