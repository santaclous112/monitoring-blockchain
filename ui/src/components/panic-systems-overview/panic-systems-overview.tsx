import {Component, h, Host, Prop, State} from '@stencil/core';
import {BaseChain, SubChain} from '../../interfaces/chains';
import {Metrics} from '../../interfaces/metrics';
import {ChainsAPI} from '../../utils/chains';
import {POLLING_FREQUENCY} from '../../utils/constants';
import {
    SystemsOverviewBaseChainNotFoundError,
    SystemsOverviewNoBaseChainSpecifiedError
} from '../../utils/errors';
import {MetricsAPI} from '../../utils/metrics';
import {PanicSystemsOverviewInterface} from './panic-systems-overview.interface';
import {SystemsOverviewAPI} from './utils/panic-systems-overview.utils';

@Component({
    tag: 'panic-systems-overview',
    styleUrl: 'panic-systems-overview.scss'
})
export class PanicSystemsOverview implements PanicSystemsOverviewInterface {

    @Prop() baseChainName: string;
    @State() systemMetrics: Metrics[];
    _chains: SubChain[] = [];
    _updater: number;
    _updateFrequency: number = POLLING_FREQUENCY;

    async componentWillLoad() {
        if (!this.baseChainName) {
            throw new SystemsOverviewNoBaseChainSpecifiedError();
        }

        try {
            // Get specified base chain.
            const baseChain: BaseChain = await ChainsAPI.getBaseChainByName(this.baseChainName);
            // Check if base chain exists.
            if (!baseChain) {
                throw new SystemsOverviewBaseChainNotFoundError(this.baseChainName);
            }

            // Store chains.
            this._chains = baseChain.subChains;
            await this.reRenderAction();

            this._updater = window.setInterval(async () => {
                await this.reRenderAction();
            }, this._updateFrequency);
        } catch (error: any) {
            console.error(error);
        }
    }

    async reRenderAction() {
        // Get metrics and alerts metrics.
        this.systemMetrics = await MetricsAPI.getMetrics(this._chains);
    }

    componentDidLoad() {
        // Update chains dropdown to show selected chain.
        const chainsDropdownLabel: Element = document.querySelector('panic-header svc-dropdown-menu ion-label');
        if (chainsDropdownLabel) {
            chainsDropdownLabel.textContent = this.baseChainName;
        }
    }

    render() {
        return (
            <Host>
                <svc-surface
                    label={`SYSTEMS OVERVIEW - ${this.baseChainName.toUpperCase()}`}
                    labelPosition={"center"}>
                    <h3 class='panic-systems-overview__subtitle'>System
                        Metrics</h3>

                    {/* Data table */}
                    {SystemsOverviewAPI.getDataTableJSX(this.systemMetrics)}

                    {SystemsOverviewAPI.checkIfAnyNotAvailable(this.systemMetrics) &&
                    <svc-label color="dark" position="start"
                               class="panic-systems-overview__info-message">N/A:
                        Data is not available from back-end API.</svc-label>}
                </svc-surface>
            </Host>
        );
    }
}
