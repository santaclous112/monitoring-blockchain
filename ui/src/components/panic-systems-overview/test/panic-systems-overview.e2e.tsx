import {newSpecPage} from '@stencil/core/testing';
import {PanicSystemsOverview} from '../panic-systems-overview';
import {h} from '@stencil/core';
import {
    fetchMock,
    mockMetricsData,
    mockMonitorablesInfoData
} from '../../../utils/mock';

fetchMock.mockResponses([JSON.stringify(mockMonitorablesInfoData())], [JSON.stringify(mockMetricsData())]);

describe('panic-systems-overview', () => {
    it('renders', async () => {
        const page = await newSpecPage({
            components: [PanicSystemsOverview],
            template: () => (<panic-systems-overview
                                baseChainName={'general'}/>),
        });

        const panicSystemsOverview = page.body.querySelector('panic-systems-overview');
        expect(panicSystemsOverview).toMatchSnapshot();
    });
});
