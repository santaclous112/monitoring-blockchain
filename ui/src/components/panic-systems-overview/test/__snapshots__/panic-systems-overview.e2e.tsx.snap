// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`panic-systems-overview renders 1`] = `
<panic-systems-overview>
  <svc-surface label="SYSTEMS OVERVIEW - GENERAL" labelposition="center">
    <h3 class="panic-systems-overview__subtitle">
      System Metrics
    </h3>
    <svc-data-table last-clicked-column-index="0" no-records-message="There are no metrics to display at this time" ordering="ascending"></svc-data-table>
  </svc-surface>
</panic-systems-overview>
`;
